### PR TITLE
Update datastore error handling for android + flutter as per #314

### PR DESCRIPTION
**Pull request by Amplifiyer**
*Description of changes:* Update error handling for android + flutter as per #314

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt [hunk 1]
```diff
@@ -110,11 +115,13 @@ class AmplifyDataStorePlugin : FlutterPlugin, MethodCallHandler {
...
-                    flutterResult,
-                    FlutterDataStoreFailureMessage.ERROR_CASTING_INPUT_IN_PLATFORM_CODE.toString(),
-                    createErrorMap(Exception(
-                            FlutterDataStoreFailureMessage.AMPLIFY_REQUEST_MALFORMED.toString())))
+            handler.post {
+                postExceptionToFlutterChannel(flutterResult, "DataStoreException",
+                        createSerializedError(ExceptionMessages.missingExceptionMessage,
+                                ExceptionMessages.missingRecoverySuggestion,
+                                "Received invalid request from Dart, modelSchemas or modelProviderVersion" +
+                                        " was not available. Request: " + request.toString()))
```
Comment (haverchuck) [inline packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt:123]: I'd suggest the following (unless I'm misunderstanding the error condition):
`"Received invalid request from Dart; modelSchemas and/or modelProviderVersion are not available.  Request: "`

**Review by haverchuck — APPROVED** [review]
I made one minor suggestion, but otherwise looks good. It's fine if you disagree with the suggestion.

File: packages/amplify_datastore_plugin_interface/lib/src/types/exception/DataStoreException.dart [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import 'package:amplify_core/types/exception/AmplifyException.dart';
+
+/// Exception thrown from DataStore Category
+class DataStoreException extends AmplifyException {
```
[the remark this answers is not in the document]
Comment (Amplifiyer) [reply packages/amplify_datastore_plugin_interface/lib/src/types/exception/DataStoreException.dart:19]: Nopes, deleting it.

File: packages/amplify_datastore/ios/Classes/FlutterDataStoreErrorHandler.swift [hunk 3]
```diff
@@ -16,82 +16,29 @@
 import Foundation
 import Amplify
 import AmplifyPlugins
+import amplify_core
 
 class FlutterDataStoreErrorHandler {
```
[the remark this answers is not in the document]
Comment (Amplifiyer) [reply packages/amplify_datastore/ios/Classes/FlutterDataStoreErrorHandler.swift:21]: Discussed this offline. iOS currently doesn't have any way to do a generic serialization for types that do not conform to Codeable which all AmplifyErrors don't

File: packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift [hunk 4]
```diff
@@ -148,15 +149,12 @@ public class SwiftAmplifyDataStorePlugin: NSObject, FlutterPlugin {
...
-            )
+                flutterResult: flutterResult)
         }
         catch {
             print("An unexpected error occured when parsing query arguments: \(error)")
-            flutterResult(FlutterDataStoreErrorHandler.createFlutterError(
-                msg: FlutterDataStoreErrorMessage.MALFORMED.rawValue,
-                errorMap: ["UNKNOWN": "\(error.localizedDescription).\nAn unrecognized error has occurred. See logs for details." ]))
+            FlutterDataStoreErrorHandler.handleDataStoreError(error: DataStoreError(error: error),
+                                                              flutterResult: flutterResult)
```
[the remark this answers is not in the document]
Comment (Amplifiyer) [reply packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift:157]: DataStoreError(error: error) would trigger a `DataStore.unknown()` Error which has [similar messages ](https://github.com/aws-amplify/amplify-ios/blob/ae75da811d11896909d82ef4157c5c8463edc897/Amplify/Categories/DataStore/DataStoreError.swift#L103-L104) which I guess is more consistent within iOS? It's a weird question to have (Strive for consistency across iOS and android for amplify-flutter generated messages or consistency across messages in iOS or android). I presume first, which is also what you are referring to?

File: packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift [hunk 5]
```diff
@@ -94,7 +94,7 @@ public class SwiftAmplifyDataStorePlugin: NSObject, FlutterPlugin {
             modelSchemas.forEach { (modelSchema) in
                 flutterModelRegistration.addModelSchema(modelName: modelSchema.name, modelSchema: modelSchema)
             }
-
```
[the remark this answers is not in the document]
Comment (Amplifiyer) [reply packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift:97]: I'm not sure what should be the right error message be in this case by quickly looking at it. Can you provide one or is this a generic unknown exception?

File: packages/amplify_datastore/ios/Classes/FlutterDataStoreErrorHandler.swift [hunk 6]
```diff
@@ -16,82 +16,29 @@
...
+                                     recoverySuggestion: error.recoverySuggestion,
+                                     underlyingError: error.underlyingError?.localizedDescription)
     }
     
-    static func createFlutterError(msg: String, errorMap: [String: Any]) -> FlutterError {
-        return FlutterError(
-                        code: "AmplifyException",
-                        message: msg,
-                        details: errorMap)
+    static func createSerializedError(message: String,
```
[the remark this answers is not in the document]
Comment (Amplifiyer) [reply packages/amplify_datastore/ios/Classes/FlutterDataStoreErrorHandler.swift:35]: `handleDataStoreError()` --calls--> `createSerializedError(error: error)` --calls--> `createSerializedError(message: String,...)`

File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt (hunk 1 above)
Comment (Amplifiyer) [reply packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt:123]: Updated.

File: packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (fjnoyp) [reply packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift:97]: This looks to be a generic unknown exception when the map passed from Dart is not correct.

File: packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (fjnoyp) [reply packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift:157]: Sounds good, that makes more sense thanks.

File: packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (fjnoyp) [reply packages/amplify_datastore/ios/Classes/SwiftAmplifyDataStorePlugin.swift:97]: This is really minor, so I'll approve now.

**Review by fjnoyp — APPROVED** [review]

**Pull request merged**